erun-ui/build.sh lint gate reports node_modules findings again when the checkout is reached through a symlink (#940 regression)

## What happened

`erun app` through the documented dev alias fails the Go lint gate in
`erun-ui/build.sh` with findings in `frontend/node_modules`, so the desktop is
never rebuilt:

```
../../erun-990/erun-ui/frontend/node_modules/flatted/golang/pkg/flatted/flatted.go:16:1: cyclomatic complexity 32 of func `Stringify` is high (> 15) (gocyclo)
../../erun-990/erun-ui/frontend/node_modules/flatted/golang/pkg/flatted/flatted.go:145:1: cyclomatic complexity 16 of func `Parse` is high (> 15) (gocyclo)
../../erun-990/erun-ui/frontend/node_modules/flatted/golang/pkg/flatted/flatted.go:35:88: inline: Constant reflect.Ptr should be inlined (govet)
../../erun-990/erun-ui/frontend/node_modules/flatted/golang/pkg/flatted/flatted.go:62:20: inline: Constant reflect.Ptr should be inlined (govet)
4 issues:
* gocyclo: 2
* govet: 2
>> skipping desktop rebuild: build.sh exited non-zero (missing wails/yarn/node toolchain?)
```

This is #940's failure, with both of #940's fixes in place. Those fixes were
(a) `linters.exclusions.paths: ^frontend/node_modules/` in `erun-ui/.golangci.yml`
and (b) `GOLANGCI_LINT_CACHE=$SCRIPT_DIR/.golangci-cache` in `erun-ui/build.sh:74`,
scoped per checkout. Neither covers this case, because there is only **one**
checkout involved — reached by two different path spellings.

### Mechanism, confirmed step by step

1. `flatted`, a transitive npm dep of the frontend, ships a Go port at
   `frontend/node_modules/flatted/golang/pkg/flatted/flatted.go`, with no
   `go.mod` of its own. It is therefore a package of the `erun-ui` module, and
   `go list ./...` walks into it — the go tool has no special case for
   `node_modules` (verified on go1.26.3: `example.com/x/frontend/node_modules/flatted/golang/pkg/flatted`).
   That is what the `.golangci.yml` exclusion exists to suppress.

2. The dev alias is `~/.cache/erun/src/current/erun-cli/run.sh` (the shape the
   `ERUN_DEV_BIN_DIR` comment in `erun-cli/run.sh:7-14` is written for), where
   `current` is a symlink to `~/.cache/erun/src/erun-990`. `SCRIPT_DIR` uses
   plain `pwd`, so it keeps the **logical** spelling; `build.sh` runs
   golangci-lint with cwd `~/.cache/erun/src/current/erun-ui`.

3. `$SCRIPT_DIR/.golangci-cache` resolves to the same physical directory either
   way, so the \"scoped\" cache is shared between the two spellings. The cache
   had been populated by an earlier run whose cwd was the physical path — cache
   entry `01/018780a5…-d` holds both the literal text
   `cyclomatic complexity 32 of func` and the absolute path
   `…/src/erun-990/erun-ui/frontend/node_modules/flatted/…/flatted.go`. Zero
   entries in that cache carry the `current` spelling.

4. Replay hands golangci-lint the physically-spelled filename, which it
   rebases against the logical cwd: `filepath.Rel(\".../src/current/erun-ui\", \".../src/erun-990/erun-ui/frontend/node_modules/…\")`
   = `../../erun-990/erun-ui/frontend/node_modules/…`. The anchored
   `^frontend/node_modules/` no longer matches, so the findings are emitted.

5. `build.sh` runs under `set -eu`, so the gate aborts before `yarn build` and
   `go build`; `erun-cli/run.sh:99` prints the misleading \"missing
   wails/yarn/node toolchain?\" line and the operator keeps running the previous
   desktop binary.

A fresh analysis is clean — the paths only go wrong on replay:

| cache | cwd spelling | result |
| --- | --- | --- |
| empty | physical | `0 issues.` |
| empty | symlink | `0 issues.` |
| warm (written from physical cwd) | symlink | the 4 issues above |

## What you expected

The gate judges the tree it is building, regardless of which spelling of that
tree's path the caller used. `node_modules` is excluded either way.

## Reproduction

Minimal, self-contained (reproduces the output byte-for-byte apart from `../` depth):

```sh
mkdir -p repro/real/frontend/node_modules/flatted/golang/pkg/flatted && cd repro/real
printf 'module example.com/erun-ui\n\ngo 1.26.0\n' > go.mod
printf 'package main\n\nfunc main() {}\n' > main.go
cp <erun>/erun-ui/frontend/node_modules/flatted/golang/pkg/flatted/flatted.go frontend/node_modules/flatted/golang/pkg/flatted/
cat > .golangci.yml <<'YML'
version: "2"
linters:
  exclusions:
    paths:
      - ^frontend/node_modules/
  enable: [govet, gocyclo]
  settings:
    gocyclo:
      min-complexity: 15
YML
cd .. && ln -s "$PWD/real" cur && mkdir cache

(cd real && GOLANGCI_LINT_CACHE=$PWD/../cache golangci-lint run ./...)   # 0 issues.
(cd cur  && GOLANGCI_LINT_CACHE=$PWD/../cache golangci-lint run ./...)   # 4 issues, ../real/frontend/node_modules/…
```

Or in situ: run the gate once with cwd `~/.cache/erun/src/erun-990/erun-ui`,
then run `erun app` through a `current -> erun-990` symlink alias.

## Suggested fix

Both halves are one-liners, and each alone is sufficient; together they also fix
the cross-checkout case #940 originally reported:

1. Unanchor the exclusions in `erun-ui/.golangci.yml` so they survive any
   rebasing — `(^|/)node_modules/` instead of `^frontend/node_modules/` and
   `^playwright/node_modules/`. Verified: with the warm, physically-spelled
   cache and the symlink cwd, this returns `0 issues.`
2. Make the cwd spelling canonical — `SCRIPT_DIR=$(CDPATH= cd -- "$(dirname "$0")" && pwd -P)`
   in `erun-ui/build.sh:6` and `erun-cli/run.sh:6`. Then the cache key spelling
   and the cwd always agree, and the `.golangci-cache` scoping comment at
   `build.sh:67-73` becomes true as written.

Worth considering separately: `erun-cli/run.sh:99` attributes any non-zero
`build.sh` exit to a missing toolchain, which sent this one in the wrong
direction. Distinguishing \"gate failed\" from \"toolchain absent\" would help.

## Environment

- erun version: erun 1.0.179 (282b0796ef70 built 2026-08-14T19:39:57Z), latest stable 1.0.178
- context: laptop (host-side orchestrator, not inside a deployed env)
- golangci-lint 2.12.2 (built with go1.26.2), go1.26.3 darwin/arm64
- found while: `erun app` silently kept running the previous desktop build

--- a/erun-cli/run.sh
+++ b/erun-cli/run.sh
@@ -3,7 +3,10 @@
 set -eu
 
 ORIGINAL_DIR=$(pwd)
-SCRIPT_DIR=$(CDPATH= cd -- "$(dirname "$0")" && pwd)
+# -P so every path derived from here has one spelling, whichever symlinked
+# route the caller reached this script through — the desktop build this
+# delegates to gates on that.
+SCRIPT_DIR=$(CDPATH= cd -- "$(dirname "$0")" && pwd -P)
 # Where this dev wrapper writes the binaries it builds. Defaults to the
 # in-repo bin/ (gitignored), which keeps `erun app`'s sibling-directory lookup
 # for erun-app/ERun.app working unchanged. ERUN_DEV_BIN_DIR moves them out of

--- a/erun-integration/internal/normalize/desktopidentity_test.go
+++ b/erun-integration/internal/normalize/desktopidentity_test.go
@@ -1,0 +1,25 @@
+package normalize
+
+import "testing"
+
+// TestDesktopIdentityPathIsHostIndependent locks that the desktop identity path
+// collapses to one token on every host. It resolves through os.UserConfigDir(),
+// which the harness cannot pin the way ERUN_HOST_OS_OVERRIDE pins a code branch:
+// darwin returns "$HOME/Library/Application Support" and Linux the XDG dir. The
+// generic temp-dir rule stops at that space, so before this rule the scenario
+// recorded one golden on darwin and a different one on Linux — and whichever
+// host did not record it went red. Drives Apply directly so it runs everywhere.
+func TestDesktopIdentityPathIsHostIndependent(t *testing.T) {
+	const want = "no desktop identity at <DESKTOP_IDENTITY>; open an environment from the ERun desktop app once\n"
+	cases := []struct{ name, path string }{
+		{"darwin user config dir", "/private/var/folders/ab/TestMCP/001/Library/Application Support/ERun/desktopid.key"},
+		{"linux xdg config dir", "/tmp/TestMCP/001/ERun/desktopid.key"},
+		{"macos temp root without the private prefix", "/var/folders/ab/TestMCP/001/Library/Application Support/ERun/desktopid.key"},
+	}
+	for _, c := range cases {
+		raw := "no desktop identity at " + c.path + "; open an environment from the ERun desktop app once\n"
+		if got := Apply(raw); got != want {
+			t.Errorf("%s:\n got: %q\nwant: %q", c.name, got, want)
+		}
+	}
+}

--- a/erun-integration/internal/normalize/normalize.go
+++ b/erun-integration/internal/normalize/normalize.go
@@ -41,6 +41,12 @@ var defaultRules = []Replacement{
 	// token — before the generic <TMP> rule below would swallow the whole path — so
 	// goldens still show that state lives off the (project <TMP>) playbook tree.
 	{regexp.MustCompile(`/(?:private/)?(?:var/folders|var/tmp|tmp)/[^\s'"]*?/\.erun/terraform`), "<STATE_ROOT>"},
+	// The desktop identity resolves through os.UserConfigDir(), which no test
+	// seam can pin: darwin puts it under "Library/Application Support" and Linux
+	// under the XDG dir. The generic rule below stops at the space, so the same
+	// scenario left a different remainder on each OS and its golden could only be
+	// green on whichever one recorded it. Collapse the whole path first.
+	{regexp.MustCompile(`/(?:private/)?(?:var/folders|var/tmp|tmp)/[^\s'"]*(?:/Library/Application Support)?/ERun/desktopid\.key`), "<DESKTOP_IDENTITY>"},
 	{regexp.MustCompile(`/(?:private/)?(?:var/folders|var/tmp|tmp)/[^\s'"]+`), "<TMP>"},
 	// A separate rule for temp paths whose separators are percent-escaped,
 	// which the plain-path rule above cannot match.

--- a/erun-integration/testdata/mcp/token_without_a_desktop_identity_says_where_it_comes_from.txt
+++ b/erun-integration/testdata/mcp/token_without_a_desktop_identity_says_where_it_comes_from.txt
@@ -1,3 +1,3 @@
 audit: erun mcp token
 mcp: team/dev edge resolved to http://<LOOPBACK>:17000/mcp
-no desktop identity at <TMP> open an environment from the ERun desktop app once so the identity exists and deployed runtimes trust it
+no desktop identity at <DESKTOP_IDENTITY>; open an environment from the ERun desktop app once so the identity exists and deployed runtimes trust it

--- a/erun-ui/.golangci.yml
+++ b/erun-ui/.golangci.yml
@@ -6,8 +6,11 @@ run:
 linters:
   exclusions:
     paths:
-      - ^frontend/node_modules/
-      - ^playwright/node_modules/
+      # Third-party Go vendored by npm sits inside this module, so `./...`
+      # analyses it. Match the segment, not an anchored prefix: a finding
+      # replayed from the content-keyed cache renders the path spelling of
+      # the run that cached it, which need not start at this working tree.
+      - (^|/)node_modules/
   enable:
     - errcheck
     - govet

--- a/erun-ui/build.sh
+++ b/erun-ui/build.sh
@@ -3,7 +3,9 @@
 set -eu
 
 ORIGINAL_DIR=$(pwd)
-SCRIPT_DIR=$(CDPATH= cd -- "$(dirname "$0")" && pwd)
+# -P so a checkout reached through a symlink resolves to the one spelling the
+# lint gate below caches its findings against.
+SCRIPT_DIR=$(CDPATH= cd -- "$(dirname "$0")" && pwd -P)
 TARGET=${1:-"$SCRIPT_DIR/bin/erun-app"}
 VERSION_FILE="$SCRIPT_DIR/../erun-devops/VERSION"
 WAILS_BIN="${WAILS_BIN:-$(go env GOPATH)/bin/wails}"
@@ -65,12 +67,12 @@ if [ -d frontend ]; then
 		# rather than in the shared `make check` the image runs.
 		#
 		# The cache is scoped to this checkout because golangci-lint keys it by
-		# module, not by working tree: a second checkout of erun-ui replays the
-		# first one's results, and those carry the *other* tree's paths. Those
-		# paths no longer match .golangci.yml's anchored node_modules
-		# exclusions, so vendored third-party Go that this gate deliberately
-		# ignores comes back as findings against a tree that does not contain
-		# it. Scoping the cache makes the gate judge what it is building.
+		# module, not by working tree: a second checkout of erun-ui would
+		# otherwise replay the first one's results against a tree that never
+		# produced them. Correctness does not rest on that alone — a replayed
+		# finding still renders the path of the run that cached it, so
+		# .golangci.yml matches excluded paths by segment instead of anchoring
+		# them to a working tree.
 		(cd "$SCRIPT_DIR" && GOLANGCI_LINT_CACHE="${GOLANGCI_LINT_CACHE:-$SCRIPT_DIR/.golangci-cache}" golangci-lint run ./...)
 	fi
 	"$YARN_BIN" build

